src/trace/log.zig

```diff
@@ -1,366 +1,463 @@
-
-pub const default_logger: *Logger = &global;
-var global: Logger = .{ .standard = undefined };
+pub const Config = struct {
+    max_level: Level = Level.debug,
+    allocator: std.mem.Allocator,
+    /// Maximum memory that logger can use.
+    max_buffer: ?u64 = null,
+};
 
-pub const Logger = union(enum) {
-    standard: *StandardErrLogger,
-    test_logger: TestLogger,
+const LogKind = enum {
+    standard,
+    test_logger,
     noop,
+};
 
-    const Self = @This();
-
-    pub fn init(allocator: std.mem.Allocator, max_level: Level) Self {
-        return .{ .standard = StandardErrLogger.init(allocator, max_level) };
-    }
-
-    pub fn spawn(self: Self) void {
-        switch (self) {
-            .standard => |logger| {
-                logger.spawn();
-            },
-            .noop, .test_logger => {},
-        }
-    }
-
-    pub fn deinit(self: Self) void {
-        switch (self) {
-            .standard => |logger| {
-                logger.deinit();
-            },
-            .noop, .test_logger => {},
-        }
-    }
-
-    pub fn field(self: Self, name: []const u8, value: anytype) Entry {
-        switch (self) {
-            inline .standard, .test_logger => |logger| {
-                return logger.field(name, value);
-            },
-            .noop => {
-                return .noop;
-            },
-        }
-    }
-
-    pub fn infof(self: Self, comptime fmt: []const u8, args: anytype) void {
-        switch (self) {
-            inline .standard, .test_logger => |logger| {
-                logger.infof(fmt, args);
-            },
-            .noop => {},
-        }
-    }
-
-    pub fn debugf(self: Self, comptime fmt: []const u8, args: anytype) void {
-        switch (self) {
-            inline .standard, .test_logger => |logger| {
-                logger.debugf(fmt, args);
-            },
-            .noop => {},
+/// A ScopedLogger could either be:
+/// - A StandardErrLogger
+/// - A TestingLogger
+pub fn ScopedLogger(comptime scope: ?[]const u8) type {
+    return union(LogKind) {
+        const Self = @This();
+        standard: *ChannelPrintLogger,
+        test_logger: *DirectPrintLogger,
+        noop: void,
+
+        /// Can be used in tests to minimize the amount of logging during tests.
+        pub const TEST_DEFAULT_LEVEL: Level = .warn;
+
+        pub fn unscoped(self: Self) Logger {
+            return switch (self) {
+                .standard => |logger| .{ .standard = logger },
+                .test_logger => |logger| .{ .test_logger = logger },
+                .noop => .noop,
+            };
         }
-    }
-
-    pub fn warnf(self: Self, comptime fmt: []const u8, args: anytype) void {
-        switch (self) {
-            inline .standard, .test_logger => |logger| {
-                logger.warnf(fmt, args);
-            },
-            .noop => {},
+        pub fn withScope(self: Self, comptime new_scope: []const u8) ScopedLogger(new_scope) {
```

Comment:
```suggestion
        }

        pub fn withScope(self: Self, comptime new_scope: []const u8) ScopedLogger(new_scope) {
```

Reply:
this commit looks a little confusing but is a simple change -- all it does is add an empty new-line between two methods - specifically, `pub fn withScope` -- if you look at the source code without the diff its much easier to see

Reply:
[cef4177](https://github.com/Syndica/sig/pull/277/commits/cef4177679fdfeedfab823689df699770d8f7563)